jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java

```diff
@@ -414,63 +465,49 @@ protected void onFillInterestedFailed(Throwable cause)
      * Extra bytes from the initial HTTP upgrade that need to
      * be processed by the websocket parser before starting
      * to read bytes from the connection
+     *
      * @param prefilled the bytes of prefilled content encountered during upgrade
      */
     protected void setInitialBuffer(ByteBuffer prefilled)
     {
         if (LOG.isDebugEnabled())
         {
-            LOG.debug("set Initial Buffer - {}",BufferUtil.toDetailString(prefilled));
+            LOG.debug("set Initial Buffer - {}", BufferUtil.toDetailString(prefilled));
         }
         prefillBuffer = prefilled;
     }
 
-    private void notifyError(Throwable t)
-    {
-        getParser().getIncomingFramesHandler().incomingError(t);
-    }
-
-    @Override
-    public void onOpen()
-    {
-        if(LOG.isDebugEnabled())
-            LOG.debug("[{}] {}.onOpened()",policy.getBehavior(),this.getClass().getSimpleName());
-        super.onOpen();
-        this.ioState.onOpened();
-    }
-
     /**
      * Event for no activity on connection (read or write)
+     *
      * @return true to signal that the endpoint must be closed, false to keep the endpoint open
      */
     @Override
     protected boolean onReadTimeout(Throwable timeout)
     {
-        IOState state = getIOState();
-        ConnectionState cstate = state.getConnectionState();
-        if (LOG.isDebugEnabled())
-            LOG.debug("{} Read Timeout - {}",policy.getBehavior(),cstate);
-
-        if (cstate == ConnectionState.CLOSED)
+        session.callApplicationOnError(timeout);
+        String reason = timeout.getMessage();
+        if (StringUtil.isBlank(reason))
         {
-            if (LOG.isDebugEnabled())
-                LOG.debug("onReadTimeout - Connection Already CLOSED");
-            // close already completed, extra timeouts not relevant
-            // allow underlying connection and endpoint to disconnect on its own
-            return true;
+            // an exception without a message.
+            reason = timeout.getClass().getSimpleName();
         }
 
-        try
+        if (connectionState.canWriteWebSocketFrames())
```

Comment:
Consider connectionState.timeout() instead

Reply:
Decided to not use connectionState.timeout() but just to forward the Throwable to close(cause);